### ISSUE TITLE
swap model to llama3/gemma/mistral

### ISSUE THREAD

**Issue by edchengg** [hitz-zentroa/GoLLIE#17]
If i want to change the base model to something else for fine tuning, what should I be aware of and modify? i see the codebase has a flash attention modeling code for llama2. i’m just curious if it would be possible to just load models from huggingface or i still need to add the flash attention code manually?

Thanks a lot!

**Comment by ikergarcia1996** [comment]
Hi @edchengg!

This codebase was developed when Flash Attention was not yet integrated into huggingface. That is why we used a custom implementation. 

If you plan to fine-tune a model, we have an improved codebase. Please use the `dev` branch of this fork: https://github.com/hitz-zentroa/This-is-not-a-Dataset/tree/dev

This codebase uses the hugging face integration of flash attention and supports MOE models, along with other small improvements. The model configuration and the way the library works are the same. We just made some changes to the  `load_model` function and the trainer. 

We are planning to release GoLLIE2 in the future using the new codebase and more data. However, both Oscar and I are currently writing our PhD theses, so it will need to wait a few more months.

If you find any issue with the updated codebase, let me know and I will try to help you :D

**Comment by edchengg** [comment]
This is fantastic! Thanks a lot and I will try the branch today.
Btw, are you aware of any easy inference speed-up tricks? It takes a lot of time to run evaluation on large dataset so I was wondering if there are any simple things I can add on to the huggingface to speed up it.
Congrats and good luck with your PhD theses :D

**Comment by ikergarcia1996** [comment]
If you are using LoRA, merge the adapters before the evaluation. It will be done automatically if you don't use 4-bit / 8-bit quantization. If you use quantization, you can set the parameter `merge_lora_before_inference: true` in your config. You can also do it manually. 
In our experiments, the merged model is up to 10 times faster during inference. 

If you are already doing inference with a merged model, you can try to use an API optimized for inference such as vLLM (https://docs.vllm.ai/en/latest/) or TGI (https://huggingface.co/docs/text-generation-inference/index). You can store the results in a fine and use our `evaluate.py` file afterwards.

**Comment by edchengg** [comment]
Thanks! The new branch works perfectly and the 'merge_lora_before_evaluation' works as well.

**Comment by edchengg** [comment]
Hi @ikergarcia1996 ,

I was able to train mistral with lora. However, when I try to run evaluation, I met this error which is very strange inside flash attention (codellama is fine).

`File "/coc/pskynet6/ychen3411/anaconda3/envs/transfusion/lib/python3.9/site-packages/flash_attn/flash_attn_interface.py", line 80, in _flash_attn_varlen_forward
    out, q, k, v, out_padded, softmax_lse, S_dmask, rng_state = flash_attn_cuda.varlen_fwd(
RuntimeError: query and key must have the same dtype`
I tried to print out the dtype of query and key:
`torch.bfloat16 torch.bfloat16
torch.bfloat16 torch.bfloat16
torch.bfloat16 torch.bfloat16
torch.bfloat16 torch.bfloat16
torch.bfloat16 torch.bfloat16
torch.bfloat16 torch.float32`

I saw the config in https://github.com/ikergarcia1996/GoLLIE/blob/dev/configs/model_configs/eval/GoLLIE-7B_Mistral_BS128_R128.yaml and there is something called bf16_full_eval

I wonder do you have any idea of this? Thanks a lot.

**Comment by ikergarcia1996** [comment]
HI @edchengg !

Yes, that is an error I have encountered multiple times. It doesn't occur when using deepspeed or bitsandbytes quantization. I believe it's a problem with Hugging Face, unrelated to our codebase. There are numerous similar issues reported for different models. The easiest fixes are disabling flash attention or enabling deepspeed. Our code simply loads the model using `dtype=torch.bfloat16` and utilizes the Hugging Face trainer for running the inference. We don't handle the inference; everything happens inside the trainer.

In the code in this repository, we solved it by manually casting all the values to the correct dtype. But in the updated branch, since we don't use a custom flash attention implementation, we cannot manually fix the problem.
